lib/gemini.c

```diff
@@ -0,0 +1,296 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) Daniel Stenberg, <daniel@haxx.se>, et al.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ * SPDX-License-Identifier: curl
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#if defined(USE_SSL) && !defined(CURL_DISABLE_GEMINI)
+
+#include <string.h>
+#include "urldata.h"
+#include "transfer.h"
+#include "connect.h"
+#include "cfilters.h"
+#include "sendf.h"
+#include "multiif.h"
+#include "select.h"
+#include "strdup.h"
+#include "url.h"
+#include "curl_printf.h"
+#include "gemini.h"
+#include "memdebug.h"
+
+static CURLcode gemini_setup_connection(struct Curl_easy *data,
+                                        struct connectdata *conn)
+{
+  (void)conn;
+  DEBUGASSERT(data->req.p.gemini == NULL);
+
+  data->req.p.gemini = calloc(1, sizeof(struct GEMINI));
+
+  if(!data->req.p.gemini)
+    return CURLE_OUT_OF_MEMORY;
+
+  return CURLE_OK;
+}
+
+static CURLcode gemini_connect(struct Curl_easy *data, bool *done)
+{
+  char *url = data->state.url;
+  struct GEMINI *gemini = data->req.p.gemini;
+
+  /* url is guaranteed to be non-NULL and non-empty */
+  DEBUGASSERT(url);
+
+  /* gemini is guaranteed to be non-NULL */
+  DEBUGASSERT(gemini);
+
+  /* url must not start with 0xFEFF */
+  if((unsigned char)url[0] == 0xFF && (unsigned char)url[1] == 0xFE) {
+    failf(data, "URL starts with 0xFEFF");
```

Comment:
This looks very specific. Shouldn't the URL be rejected already in the URL parser if this is an invalid UR? Why would a URL start with this byte combo and why is it so explicitly bad? Where is the spec for the gemini URL format?

Reply:
It is a specific requirement that a response should not begin with a Unicode byte-order mark—uncommon but allowed for UTF-8, and IIRC very early-on it was a bug in a few Windows clients using a UTF-8 encoder with the default configuration. I think it *should* be caught as in valid by the URL parser but I suppose @jecxjo wanted to make it explicit. 

> <URL> is a UTF-8 encoded absolute URL, including a scheme, of maximum length 1024 bytes. The request MUST NOT begin with a U+FEFF byte order mark.
— [Gemini specification](https://gemini.circumlunar.space/docs/specification.gmi) § 2

Reply:
A URL (as far as curl is concerned) starts with a scheme. A scheme can only be, in the words of RFC 3986:
~~~
scheme      = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )